src/coreclr/src/jit/flowgraph.cpp

```diff
@@ -14549,37 +14561,66 @@ bool Compiler::fgBlockEndFavorsTailDuplication(BasicBlock* block)
     // is an assignment of a constant, arraylength, or a relop.
     // This is because these statements produce information about values
     // that would otherwise be lost at the upcoming merge point.
-    GenTree* tree = lastStmt->GetRootNode();
-    if (tree->gtOper != GT_ASG)
-    {
-        return false;
-    }
+    //
+    // Check up to N statements...
+    //
+    const int  limit = 2;
+    int        count = 0;
+    Statement* stmt  = lastStmt;
 
-    if (tree->OperIsBlkOp())
+    while (count < limit)
     {
-        return false;
-    }
+        count++;
+        GenTree* const tree = stmt->GetRootNode();
+        if (tree->OperIs(GT_ASG) && !tree->OperIsBlkOp())
+        {
+            GenTree* const op1 = tree->AsOp()->gtOp1;
 
-    GenTree* op2 = tree->AsOp()->gtOp2;
-    if (op2->gtOper != GT_ARR_LENGTH && !op2->OperIsConst() && ((op2->OperKind() & GTK_RELOP) == 0))
-    {
-        return false;
+            if (op1->IsLocal())
+            {
+                const unsigned op1LclNum = op1->AsLclVarCommon()->GetLclNum();
+
+                if (op1LclNum == lclNum)
+                {
+                    GenTree* const op2 = tree->AsOp()->gtOp2;
+
+                    if (op2->OperIs(GT_ARR_LENGTH) || op2->OperIsConst() || op2->OperIsCompare())
+                    {
+                        return true;
+                    }
+                }
+            }
+        }
+
+        Statement* const prevStmt = stmt->GetPrevStmt();
+
+        if (prevStmt == lastStmt)
```

Comment:
It would be worth a comment here that `GetPrevStmt()` of the fist stmt returns the last stmt. I had forgotten and was initially confused.

Reply:
Yes, this confuses me at times too. I suppose it saves a pointer in `BasicBlock` but it makes reverse iteration through statements a bit clunky.

Seems like we could have `GetPrevStmt()` return null for the first statement, and then have `BasicBlock::LastStatement()` call some other API or have friend access to the underlying field.